File: typescript/packages/mechanisms/evm/src/batch-settlement/facilitator/deposit-eip3009.ts
import { PaymentRequirements, VerifyResponse } from "@x402/core/types";
import { getAddress, isAddressEqual, parseErc6492Signature } from "viem";
import { FacilitatorEvmSigner } from "../../signer";
import { BatchSettlementDepositPayload } from "../types";
import { ERC3009_DEPOSIT_COLLECTOR_ADDRESS, receiveAuthorizationTypes } from "../constants";
import { buildErc3009CollectorData, buildErc3009DepositNonce } from "../encoding";
import * as Errors from "../errors";
import { erc3009AuthorizationTimeInvalidReason } from "./utils";
import { verifyTypedDataSignature } from "../../shared/verifySignature";

const ZERO_ADDRESS = "0x0000000000000000000000000000000000000000" as const;

/** Factory deployment info for an undeployed (ERC-6492 counterfactual) deposit wallet. */
export type Erc3009CounterfactualDeployment = {
  factory: `0x${string}`;
  factoryCalldata: `0x${string}`;
};

/**
 * Result of ERC-3009 deposit authorization verification.
 *
 * - `response` non-null: terminal verification outcome (a rejection).
 * - `counterfactual` non-null: the deposit is from an undeployed ERC-6492 wallet with an
 *   allowlisted factory; the caller must validate the inner signature via the deploy+deposit
 *   simulation rather than a direct (no-code) signature check.
 * - both null: a deployed wallet / plain EOA signature is valid.
 */
export type Erc3009DepositVerifyResult = {
  response: VerifyResponse | null;
  counterfactual: Erc3009CounterfactualDeployment | null;
};

/**
 * Returns the collector contract used for EIP-3009 deposits.
 *
 * @returns ERC-3009 deposit collector address.
 */
export function getEip3009DepositCollectorAddress(): `0x${string}` {
  return getAddress(ERC3009_DEPOSIT_COLLECTOR_ADDRESS);
}

/**
 * Encodes collector data for an EIP-3009 deposit payload.
 *
 * @param payload - Deposit payload containing the ERC-3009 authorization.
 * @returns ABI-encoded collector data.
 */
export function buildEip3009DepositCollectorData(
  payload: BatchSettlementDepositPayload,
): `0x${string}` {
  const auth = payload.deposit.authorization.erc3009Authorization;
  if (!auth) {
    throw new Error(Errors.ErrErc3009AuthorizationRequired);
  }

  const { signature } = parseErc6492Signature(auth.signature);
  return buildErc3009CollectorData(auth.validAfter, auth.validBefore, auth.salt, signature);
}

/**
 * Verifies the ERC-3009 authorization fields and typed-data signature.
 *
 * @param signer - Facilitator signer for typed-data verification.
 * @param payload - Deposit payload to verify.
 * @param requirements - Payment requirements containing token domain metadata.
 * @param chainId - EVM chain id.
 * @param allowedFactories - Allowlisted ERC-6492 factory addresses for counterfactual deposits.
 * @returns The verification result (rejection, valid, or counterfactual-deferred).
 */
export async function verifyEip3009DepositAuthorization(
  signer: FacilitatorEvmSigner,
  payload: BatchSettlementDepositPayload,
  requirements: PaymentRequirements,
  chainId: number,
  allowedFactories: string[] = [],
): Promise<Erc3009DepositVerifyResult> {
  const { deposit, voucher } = payload;
  const payer = payload.channelConfig.payer;
  const auth = deposit.authorization.erc3009Authorization;

  const reject = (invalidReason: string): Erc3009DepositVerifyResult => ({
    response: { isValid: false, invalidReason, payer },
    counterfactual: null,
  });

  if (!auth) {
    return reject(Errors.ErrErc3009AuthorizationRequired);
  }

  const extra = requirements.extra as { name?: string; version?: string } | undefined;
  if (!extra?.name || !extra?.version) {
    return reject(Errors.ErrMissingEip712Domain);
  }

  const validAfter = BigInt(auth.validAfter);
  const validBefore = BigInt(auth.validBefore);
  const timeInvalid = erc3009AuthorizationTimeInvalidReason(validAfter, validBefore);
  if (timeInvalid) {
    return reject(timeInvalid);
  }

  // Parse the ERC-6492 wrapper (a no-op for unwrapped signatures, which return the signature
  // unchanged as `signature`).
  const {
    address: factory,
    data: factoryCalldata,
    signature: innerSig,
  } = parseErc6492Signature(auth.signature);
  const hasDeploymentInfo = !!(
    factory &&
    factoryCalldata &&
    !isAddressEqual(factory, ZERO_ADDRESS)
  );

  // Counterfactual detection: only fetch code when there is deployment info so the common
  // (already-deployed / plain EOA) path keeps a single RPC round-trip.
  if (hasDeploymentInfo) {
    let code: `0x${string}` | undefined;
    try {
      code = await signer.getCode({ address: payer });
    } catch {
      code = undefined;
    }
    if (!code || code === "0x") {
      const normalizedFactory = factory.toLowerCase();
      const isAllowed = allowedFactories.some(a => a.trim().toLowerCase() === normalizedFactory);
      if (!isAllowed) {
        return reject(Errors.ErrFactoryNotAllowed);
      }
      // Counterfactual + allowlisted: defer signature validation to the deploy+deposit
      // simulation performed by the caller.
      return {
        response: null,
        counterfactual: { factory, factoryCalldata: factoryCalldata as `0x${string}` },
      };
    }
    // Already deployed despite the wrapper — fall through and validate the inner signature.
  }

  const erc3009Nonce = buildErc3009DepositNonce(voucher.channelId, auth.salt);
  const receiveAuthOk = await verifyReceiveAuth(signer, {
    payer,
    asset: requirements.asset,
    name: extra.name,
    version: extra.version,
    chainId,
    amount: deposit.amount,
    validAfter,
    validBefore,
    nonce: erc3009Nonce,
    signature: innerSig,
  });

  if (!receiveAuthOk) {
    return reject(Errors.ErrInvalidReceiveAuthorizationSignature);
  }

  return { response: null, counterfactual: null };
}

/**
 * Verifies a `ReceiveWithAuthorization` signature.
 *
 * @param signer - Facilitator signer used for typed-data verification.
 * @param params - Authorization fields and signature.
 * @param params.payer - Expected authorization signer.
 * @param params.asset - ERC-20 verifying contract.
 * @param params.name - ERC-20 EIP-712 domain name.
 * @param params.version - ERC-20 EIP-712 domain version.
 * @param params.chainId - EVM chain id.
 * @param params.amount - Authorized token amount.
 * @param params.validAfter - Earliest valid timestamp.
 * @param params.validBefore - Expiration timestamp.
 * @param params.nonce - ERC-3009 nonce.
 * @param params.signature - Receive authorization signature.
 * @returns True when the signature matches the expected payer.
 */
async function verifyReceiveAuth(
  signer: FacilitatorEvmSigner,
  params: {
    payer: `0x${string}`;
    asset: string;
    name: string;
    version: string;
    chainId: number;
    amount: string;
    validAfter: bigint;
    validBefore: bigint;
    nonce: `0x${string}`;
    signature: `0x${string}`;
  },
): Promise<boolean> {
  // Mirror the token's on-chain ERC-3009 signature check. Modern tokens (USDC v2.2)
  // use a SignatureChecker that routes by code.length: ECDSA for EOAs, strict
  // EIP-1271 for any address with code (including 7702-delegated EOAs). No
  // ECDSA fallback for code addresses — that fallback would accept sigs the
  // token rejects on-chain.
  return verifyTypedDataSignature(signer, {
    address: getAddress(params.payer),
    domain: {
      name: params.name,
      version: params.version,
      chainId: params.chainId,
      verifyingContract: getAddress(params.asset),
    },
    types: receiveAuthorizationTypes,
    primaryType: "ReceiveWithAuthorization",
    message: {
      from: getAddress(params.payer),
      to: getAddress(ERC3009_DEPOSIT_COLLECTOR_ADDRESS),
      value: BigInt(params.amount),
      validAfter: params.validAfter,
      validBefore: params.validBefore,
      nonce: params.nonce,
    },
    signature: params.signature,
  });
}

File: python/x402/mechanisms/evm/batch_settlement/facilitator/deposit_eip3009.py
"""Facilitator-side ERC-3009 deposit helpers."""

from __future__ import annotations

try:
    from eth_utils import to_checksum_address
except ImportError as e:
    raise ImportError(
        "EVM mechanism requires ethereum packages. Install with: pip install x402[evm]"
    ) from e

from .....schemas import PaymentRequirements, VerifyResponse
from ...erc6492 import has_deployment_info, parse_erc6492_signature
from ...signer import FacilitatorEvmSigner
from ...types import ERC6492SignatureData
from ...utils import bytes_to_hex
from ...verify import verify_typed_data_strict
from ..constants import (
    ERC3009_DEPOSIT_COLLECTOR_ADDRESS,
    RECEIVE_AUTHORIZATION_TYPES,
)
from ..encoding import build_erc3009_collector_data, build_erc3009_deposit_nonce
from ..errors import (
    ERR_ERC3009_AUTHORIZATION_REQUIRED,
    ERR_FACTORY_NOT_ALLOWED,
    ERR_INVALID_RECEIVE_AUTHORIZATION_SIGNATURE,
    ERR_MISSING_EIP712_DOMAIN,
)
from ..types import DepositPayload
from ..utils import coerce_bytes32
from .utils import erc3009_authorization_time_invalid_reason


def get_eip3009_deposit_collector_address() -> str:
    return to_checksum_address(ERC3009_DEPOSIT_COLLECTOR_ADDRESS)


def build_eip3009_deposit_collector_data(payload: DepositPayload) -> bytes:
    """ABI-encode collector data, stripping any ERC-6492 wrapping from the signature."""
    assert payload.deposit is not None
    auth = payload.deposit.authorization.erc3009_authorization
    if auth is None:
        raise ValueError(ERR_ERC3009_AUTHORIZATION_REQUIRED)

    sig_bytes = bytes.fromhex(auth.signature.removeprefix("0x"))
    parsed = parse_erc6492_signature(sig_bytes)
    inner_hex = "0x" + parsed.inner_signature.hex()

    return build_erc3009_collector_data(auth.valid_after, auth.valid_before, auth.salt, inner_hex)


def verify_eip3009_deposit_authorization(
    signer: FacilitatorEvmSigner,
    payload: DepositPayload,
    requirements: PaymentRequirements,
    chain_id: int,
    allowed_factories: list[str] | None = None,
) -> tuple[ERC6492SignatureData | None, VerifyResponse | None]:
    """Validate ERC-3009 timing + typed-data signature.

    Returns ``(counterfactual_sig_data, None)`` when the deposit is from an undeployed
    ERC-6492 wallet whose factory is allowlisted — the caller must then validate the inner
    signature via the deploy+deposit simulation (the wallet has no code yet, so a direct
    signature check cannot succeed). Returns ``(None, None)`` when a deployed wallet / plain
    EOA signature is valid, and ``(None, VerifyResponse)`` on any rejection.
    """
    assert payload.deposit is not None and payload.voucher is not None
    assert payload.channel_config is not None
    deposit = payload.deposit
    voucher = payload.voucher
    payer = payload.channel_config.payer
    auth = deposit.authorization.erc3009_authorization

    if auth is None:
        return None, VerifyResponse(
            is_valid=False, invalid_reason=ERR_ERC3009_AUTHORIZATION_REQUIRED, payer=payer
        )

    extra = requirements.extra or {}
    name = extra.get("name")
    version = extra.get("version")
    if not name or not version:
        return None, VerifyResponse(
            is_valid=False, invalid_reason=ERR_MISSING_EIP712_DOMAIN, payer=payer
        )

    time_invalid = erc3009_authorization_time_invalid_reason(
        int(auth.valid_after), int(auth.valid_before)
    )
    if time_invalid:
        return None, VerifyResponse(is_valid=False, invalid_reason=time_invalid, payer=payer)

    try:
        sig_data = parse_erc6492_signature(bytes.fromhex(auth.signature.removeprefix("0x")))
    except Exception:
        return None, VerifyResponse(
            is_valid=False,
            invalid_reason=ERR_INVALID_RECEIVE_AUTHORIZATION_SIGNATURE,
            payer=payer,
        )

    # Counterfactual detection: only fetch code when there is deployment info so the common
    # (already-deployed / plain EOA) path keeps a single RPC round-trip.
    if has_deployment_info(sig_data):
        code = signer.get_code(to_checksum_address(payer))
        if len(code) == 0:
            allowed = [f.strip().lower() for f in (allowed_factories or [])]
            if bytes_to_hex(sig_data.factory).lower() not in allowed:
                return None, VerifyResponse(
                    is_valid=False, invalid_reason=ERR_FACTORY_NOT_ALLOWED, payer=payer
                )
            # Counterfactual + allowlisted: defer signature validation to the
            # deploy+deposit simulation performed by the caller.
            return sig_data, None
        # Already deployed despite the wrapper — fall through and validate the inner
        # signature against its EIP-1271 validator like any other deployed wallet.

    erc3009_nonce = build_erc3009_deposit_nonce(voucher.channel_id, auth.salt)
    ok = _verify_receive_auth(
        signer,
        payer=payer,
        asset=requirements.asset,
        name=name,
        version=version,
        chain_id=chain_id,
        amount=deposit.amount,
        valid_after=int(auth.valid_after),
        valid_before=int(auth.valid_before),
        nonce=erc3009_nonce,
        signature="0x" + sig_data.inner_signature.hex(),
    )
    if not ok:
        return None, VerifyResponse(
            is_valid=False,
            invalid_reason=ERR_INVALID_RECEIVE_AUTHORIZATION_SIGNATURE,
            payer=payer,
        )

    return None, None


def _verify_receive_auth(
    signer: FacilitatorEvmSigner,
    *,
    payer: str,
    asset: str,
    name: str,
    version: str,
    chain_id: int,
    amount: str,
    valid_after: int,
    valid_before: int,
    nonce: str,
    signature: str,
) -> bool:
    try:
        sig_bytes = bytes.fromhex(signature.removeprefix("0x"))
        domain = {
            "name": name,
            "version": version,
            "chainId": chain_id,
            "verifyingContract": to_checksum_address(asset),
        }
        message = {
            "from": to_checksum_address(payer),
            "to": to_checksum_address(ERC3009_DEPOSIT_COLLECTOR_ADDRESS),
            "value": int(amount),
            "validAfter": valid_after,
            "validBefore": valid_before,
            "nonce": coerce_bytes32(nonce),
        }
        # Uses the strict primitive that mirrors on-chain SignatureChecker (code-routed, no ECDSA fallback).
        return verify_typed_data_strict(
            signer,
            address=to_checksum_address(payer),
            domain=domain,
            types=RECEIVE_AUTHORIZATION_TYPES,
            primary_type="ReceiveWithAuthorization",
            message=message,
            signature=sig_bytes,
        )
    except Exception:
        return False


__all__ = [
    "get_eip3009_deposit_collector_address",
    "build_eip3009_deposit_collector_data",
    "verify_eip3009_deposit_authorization",
]
